signer/src/bitcoin/utxo.rs

```diff
@@ -1466,6 +1482,104 @@ pub trait TxDeconstructor: BitcoinInputsOutputs {
             .collect()
     }
 
+    /// Return the withdrawal outputs, matching the tx outputs to the decoded
+    /// withdrawal IDs
+    fn to_withdrawal_outputs(
+        &self,
+        tx_outputs: &[TxOutput],
+    ) -> Result<Vec<WithdrawalTxOutput>, Error> {
+        // If the first output is not a SignersOutput, nothing to do
+        match tx_outputs.first() {
+            Some(TxOutput {
+                output_type: TxOutputType::SignersOutput,
+                ..
+            }) => (),
+            _ => return Ok(Vec::new()),
+        }
+
+        // If the second output is not a SignersOpReturn, nothing to do
+        let op_return_output = match tx_outputs.get(1) {
+            Some(
+                op_return_output @ TxOutput {
+                    output_type: TxOutputType::SignersOpReturn,
+                    ..
+                },
+            ) => op_return_output,
+            _ => return Ok(Vec::new()),
+        };
+
+        // If there are no withdrawals, nothing to do
+        if tx_outputs.len() == 2 {
+            return Ok(Vec::new());
+        }
+
+        // SAFETY: we checked that we have at least two outputs in the matches
+        let tx_withdrawals_outputs = &tx_outputs[2..];
+
+        // Sanity check: all the other outputs must be withdrawals
+        let is_all_withdrawals = tx_withdrawals_outputs
+            .iter()
+            .all(|out| out.output_type == TxOutputType::Withdrawal);
+        if !is_all_withdrawals {
+            return Err(Error::SbtcTxMalformed);
+        }
+
+        let op_return_instructions: Vec<_> = op_return_output
+            .script_pubkey
+            .as_script()
+            .instructions()
+            .collect();
+
+        // The op return script must be a OP_RETURN and a push bytes
+        let [Ok(Instruction::Op(OP_RETURN)), Ok(Instruction::PushBytes(push_bytes))] =
+            op_return_instructions[..]
+        else {
+            return Err(Error::SbtcTxOpReturnFormatError);
+        };
+
+        let raw_bytes = push_bytes.as_bytes();
+        if raw_bytes.len() < OP_RETURN_HEADER_SIZE {
+            return Err(Error::SbtcTxOpReturnFormatError);
+        }
+
+        // First two bytes are magic bytes, we don't care about them.
+        // The third one is the version byte.
+        // SAFETY: 2 < OP_RETURN_HEADER_SIZE (3)
+        let version = raw_bytes[2];
+
+        if version == 0 {
+            // In version 0 we didn't store withdrawal ids
+            return Ok(Vec::new());
+        } else if version != OP_RETURN_VERSION {
+            // Unknown version byte
+            return Err(Error::SbtcTxOpReturnFormatError);
+        }
+
+        // SAFETY: 3 <= OP_RETURN_HEADER_SIZE (3), if the encoded is empty we
```

Comment:
Maybe slightly misleading, since 3 is beyond 0,1,2 (len >= 3) we checked above. Perhaps just state that we've ensured that the slice isn't empty above and thus this will produce an empty slice rather than panicking?

Reply:
mmm wait not really: we didn't check that the slice was not empty, and `3 <= OP_RETURN_HEADER_SIZE (3)` is safe as `&raw_bytes[3..]` would return an empty slice for empty encoded withdrawal ids. Am I missing something?

Reply:
I was referring to the `SAFETY: 3 <= OP_RETURN_HEADER_SIZE (3)` statement, as `len()` is 1-indexed while the slice access is 0-indexed. So the safety lies rather in the fact that we checked `raw_bytes.len() < OP_RETURN_HEADER_SIZE` above (i.e. not empty), and the slice behavior guarantees that we'll get an empty slice when accessing index `3` instead of a panic.